### ISSUE TITLE
[Bug] logging api used inside the signal handler

### ISSUE THREAD

**Issue by KarthikNedunchezhiyan** [firecracker-microvm/firecracker#2452]
# Describe the bug

As logging API will use lock internally, I suspect using logging API inside the signal handler is not async safe. Let say logging API `error!` implementation is like below

```
error! => {
   get-lock
   write-to-sink
   release-lock
}
```

When the firecracker thread1's execution is at `get-lock` phase and suddenly signal handler is triggered, Signal handler will also try to get lock on `get-lock` phase (because the signal handler is using `error!`). Now both of them will be in a dead-lock. thread1 will wait for the signal handler to complete and the signal handler will wait for thread1 to release the lock.

## To Reproduce

It is hard to reproduce.

## Additional context

I can notice rust-sys-util blocked the other signals while the current signal handler is in use, but IMO this will not solve this problem.

I am newbie to this topic. feel free to close this if it is not valid 🙂.

## Checks

- [ ] Have you searched the Firecracker Issues database for similar problems?
- [ ] Have you read the existing relevant Firecracker documentation?
- [ ] Are you certain the bug being reported is a Firecracker issue?

**Comment by alindima** [comment]
Hi. Thank you for logging the issue.
We've looked over the potential bug and it seems that it is indeed a very hard to reproduce bug.

This is also the case for metrics, not just for logging, and it seems to go beyond mutex locks (https://man7.org/linux/man-pages/man7/signal-safety.7.html)

The fix would be to use only async-signal-safe functions in the signal handler.

**Comment by alindima** [comment]
Off the top of my head, a possible fix would be to remove the `exit` call from the signal handler, as well as the logging and metrics calls, and replace them with a notification mechanism (an eventfd maybe) that is checked regularly by the emulation thread.

**Comment by KarthikNedunchezhiyan** [comment]
Using eventfd to offload the non-async signal task from the signal handler is a common way but one challenge in offloading the task to another handler is to pass sufficient information(like si_signo, si_code) to another handler.

**Comment by alindima** [comment]
Fixed in #2459

**Comment by alindima** [comment]
We've synced internally, and discussed the potential impact of this bug, as well as the alternatives for mitigation.

We converged on the fact that the bug is very hard to reproduce (and has no impact that we know of), and that the worst case scenario is a Firecracker process deadlock, with no security implications.
We think that the possible solutions we explored are too complicated for the moment.

We decided to add a prod host setup update that states this signal handler limitation and recommends to have an overseer process that terminates Firecracker processes that are stuck in a deadlock.
